### PR TITLE
- adding "valueFromKey" method to EvolvPredicatesImpl.java

### PR DESCRIPTION
**Pull request by CreatingTheNewWorld**
- minor changes in the "Predicate" class
- adding unit tests EvolvPredicatesImplTest
- test methods : "testValueFromKey", "testEvaluateFilter", "testEvaluateRule"

### CODE REVIEW

File: android-sdk/src/main/java/ai/evolv/android_sdk/EvolvPredicatesImpl.java [hunk 1]
```diff
@@ -35,23 +39,14 @@ public JsonElement evaluate(JsonElement context, JsonElement predicate) {
...
-                if(result.has("touched")){
-                    result.get("touched").add(jsonObject.get("field"));
-                }
-            }
-            if(jsonObject.has("failed")){
-                if(result.has("touched")){
-                    result.get("touched").add(jsonObject.get("field"));
-                }
-            }
+        for (Map.Entry<String,JsonElement> entry : passedObject.entrySet()) {
```
Comment (robertsevernsentient) [inline android-sdk/src/main/java/ai/evolv/android_sdk/EvolvPredicatesImpl.java:45]: What is 'touched' for? It's not the old SDK method of seeing if the key has been requested to define if we're going to send a confirmation is it?

File: android-sdk/src/main/java/ai/evolv/android_sdk/EvolvPredicatesImpl.java [hunk 2]
```diff
@@ -90,79 +87,95 @@ private boolean evaluatePredicate(JsonElement user,
...
     }
 
-    // TODO: 11.06.2021 need a unit test
-    private boolean evaluateFilter(JsonElement user, JsonObject rule) {
-        JsonElement value = valueFromKey(user, rule.get("field").getAsJsonObject());
+     boolean evaluateFilter(JsonElement user, JsonObject rule) {
+        JsonElement value = valueFromKey(user, rule.get("field").getAsString());
 
-        if (rule.get("operator").getAsString().startsWith("kv_") && value.isJsonNull() ) {
+        if (rule.get("operator").getAsString().startsWith("kv_") || value.isJsonNull()) {
```
Comment (robertsevernsentient) [inline android-sdk/src/main/java/ai/evolv/android_sdk/EvolvPredicatesImpl.java:127]: We do not use 'kv_' operators any more

File: android-sdk/src/test/java/ai/evolv/android_sdk/EvolvPredicatesImplTest.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        //operator - "not_equal"; condition: context - "signedin":"yes", rule - "signedin":"yes"
+        String rawRule_not_equal = "{\"field\":\"signedin\",\"operator\":\"not_equal\",\"value\":\"yes\"}";
+        //operator - "contains"; condition: context - "signedin":"yes", rule - "signedin":"yes"
+        String rawRule_contains = "{\"field\":\"signedin\",\"operator\":\"contains\",\"value\":\"yes\"}";
+        //operator - "exists"; condition: context - "signedin":"yes", rule - "signedin":"yes"
+        String rawRule_exists = "{\"field\":\"signedin\",\"operator\":\"exists\",\"value\":\"yes\"}";
+        //operator - "not_contains"; condition: context - "signedin":"yes", rule - "signedin":"yes"
+        String rawRule_not_contains = "{\"field\":\"signedin\",\"operator\":\"not_contains\",\"value\":\"yes\"}";
+        //operator - "not_defined"; condition: context - "signedin":"yes", rule - "signedin":"yes"
+        String rawRule_not_defined = "{\"field\":\"signedin\",\"operator\":\"not_defined\",\"value\":\"yes\"}";
```
Comment (robertsevernsentient) [inline android-sdk/src/test/java/ai/evolv/android_sdk/EvolvPredicatesImplTest.java:84]: missing defined and not exists?

File: android-sdk/src/test/java/ai/evolv/android_sdk/EvolvPredicatesImplTest.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        Assert.assertEquals(element.getAsJsonArray().get(0).getAsJsonObject().get("eid").getAsString(), "47d857cd5e");
+        Assert.assertEquals(element.getAsJsonArray().get(0).getAsJsonObject().get("cid").getAsString(), "5fa0fd38aae6:47d857cd5e");
+        Assert.assertEquals(element.getAsJsonArray().get(0).getAsJsonObject().get("ordinal").getAsString(), "0");
+        Assert.assertEquals(element.getAsJsonArray().get(0).getAsJsonObject().get("group_id").getAsString(), "511ce252-92b5-4611-a00c-0e4120369c96");
+        Assert.assertFalse(element.getAsJsonArray().get(0).getAsJsonObject().get("excluded").getAsBoolean());
+
+    }
+
+    @Test
+    public void testEvaluateFilter() {
```
Comment (robertsevernsentient) [inline android-sdk/src/test/java/ai/evolv/android_sdk/EvolvPredicatesImplTest.java:71]: Can we add some more complex conditions with AND/OR and nested groups?

**Review by robertsevernsentient — CHANGES_REQUESTED** [review]

File: android-sdk/src/main/java/ai/evolv/android_sdk/EvolvPredicatesImpl.java (hunk 1 above)
Comment (CreatingTheNewWorld) [reply android-sdk/src/main/java/ai/evolv/android_sdk/EvolvPredicatesImpl.java:45]: изображение
изображение

1) it adds fields from both passed keys and failed keys to Set (Predicates.js)
2) copy it to "touchedKeys" set (Store.js)

File: android-sdk/src/main/java/ai/evolv/android_sdk/EvolvPredicatesImpl.java (hunk 2 above)
Comment (CreatingTheNewWorld) [reply android-sdk/src/main/java/ai/evolv/android_sdk/EvolvPredicatesImpl.java:127]: yes thanks, at that moment I did not know that they were not used 
i'll fix it

File: android-sdk/src/main/java/ai/evolv/android_sdk/EvolvStoreImpl.java [hunk 5]
```diff
@@ -243,35 +241,169 @@ private void reevaluateContext() {
...
+    // TODO: 30.06.2021 need a unit test
     private void setActiveAndEntryKeyStates(int version,
                                             EvolvContext evolvContext,
                                             JsonObject config,
                                             JsonArray allocations,
                                             KeyStates configKeyStates) {
-        // TODO: 01.06.2021 implement
 
-        Object results = evaluatePredicates(version, evolvContext, config);
+        JsonArray results = evaluatePredicates(version, evolvContext, config);
```
Comment (robertsevernsentient) [inline android-sdk/src/main/java/ai/evolv/android_sdk/EvolvStoreImpl.java:313]: These types of changes need tests when you submit PRs for them

**Review by robertsevernsentient — CHANGES_REQUESTED** [review]
We need tests included in the PR for changes like this

File: android-sdk/src/main/java/ai/evolv/android_sdk/EvolvPredicatesImpl.java (hunk 2 above)
Comment (CreatingTheNewWorld) [reply android-sdk/src/main/java/ai/evolv/android_sdk/EvolvPredicatesImpl.java]: fixed

File: android-sdk/src/main/java/ai/evolv/android_sdk/EvolvStoreImpl.java (hunk 5 above)
Comment (CreatingTheNewWorld) [reply android-sdk/src/main/java/ai/evolv/android_sdk/EvolvStoreImpl.java:376]: done

**Review by robertsevernsentient — APPROVED** [review]

**Pull request merged**